pkg/autoscaler/controller/linearcontroller/linear_controller_test.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package linearcontroller
+
+import (
+	"testing"
+
+	"github.com/davecgh/go-spew/spew"
+)
+
+func verifyParams(t *testing.T, scalerParams, expScalerParams *linearParams) {
+	if scalerParams.CoresPerReplica != expScalerParams.CoresPerReplica ||
+		scalerParams.NodesPerReplica != expScalerParams.NodesPerReplica ||
+		scalerParams.Min != expScalerParams.Min ||
+		scalerParams.Max != expScalerParams.Max {
+		t.Errorf("parser error - Expected params %v MISMATCHED: Got %v", expScalerParams, scalerParams)
+	}
+}
+
+func TestControllerParser(t *testing.T) {
+	testCases := []struct {
+		jsonData  string
+		expError  bool
+		expParams *linearParams
+	}{
+		{
+			`{ 
+		      "coresPerReplica": 2,
+		      "nodesPerReplica": 1,
+		      "min": 1,
+		      "max": 100
+		    }`,
+			false,
+			&linearParams{
+				CoresPerReplica: 2,
+				NodesPerReplica: 1,
+				Min:             1,
+				Max:             100,
+			},
+		},
+		{ // Invalid JSON
+			`{ "coresPerReplica": {{ 1:1 } }`,
+			true,
+			&linearParams{},
+		},
+		{ // Invalid string value
+			`{ "coresPerReplica": "whatisthis"`,
+			true,
+			&linearParams{},
+		},
+		{ // Invalid negative value
+			`{ "nodesPerReplica":  -20 }`,
+			true,
+			&linearParams{},
+		},
+		{ // Invalid max that smaller tham min
+			`{ 
+		      "nodesPerReplica": 1,
+		      "min": 100,
+		      "max": 50
+		    }`,
+			true,
+			&linearParams{},
+		},
+		{ // Both coresPerReplica and nodesPerReplica are unset
+			`{ 
+		      "min": 1,
+		      "max": 100
+		    }`,
+			true,
+			&linearParams{},
+		},
+	}
+
+	for _, tc := range testCases {
+		params, err := parseParams([]byte(tc.jsonData))
+		if tc.expError {
+			if err == nil {
+				t.Errorf("unexpected parsing success. Expected failure")
+				spew.Dump(tc)
+				spew.Dump(params)
+			}
+			continue
+		}
+		if err != nil && !tc.expError {
+			t.Errorf("unexpected parse failure")
+			spew.Dump(tc)
+			continue
+		}
+		verifyParams(t, params, tc.expParams)
+	}
+}
+
+func TestControllerScaler(t *testing.T) {
+	testController := &LinearController{}
+	testController.params = &linearParams{
+		CoresPerReplica: 2,
+		Min:             2,
+		Max:             100,
+	}
+
+	testCases := []struct {
+		controller   *LinearController
+		numResources int
+		expReplicas  int
+	}{
+		{testController, 0, 2},
```

Comment:
Why is testController the same for all the testcases? can we remove this field from the struct?